mountpoint-s3-crt/src/io/futures.rs

```diff
@@ -220,17 +226,122 @@ pub enum JoinError {
+        // SAFETY: `self.inner` contains a valid `aws_future_void`.
+        unsafe { aws_future_void_is_done(self.inner.as_ptr()) }
+    }
+
+    /// Create a [FutureVoid] from a [aws_future_void].
+    ///
+    /// ## Safety
+    ///
+    /// `inner` must be a valid [aws_future_void] with no registered callbacks.
+    pub unsafe fn from_crt(inner: NonNull<aws_future_void>) -> Self {
+        Self {
+            inner,
+            waker: Arc::new(Mutex::new(None)),
+        }
+    }
+
+    /// Get the result of this future if completed.
+    fn try_get_result(&self) -> Option<Result<(), crate::common::error::Error>> {
+        if !self.is_done() {
+            return None;
+        }
+
+        let result = {
+            // SAFETY: `self.inner` has completed.
+            let future_result = unsafe { aws_future_void_get_error(self.inner.as_ptr()) };
+            let error_result: crate::common::error::Error = future_result.into();
+            if error_result.is_err() {
+                Err(error_result)
+            } else {
+                Ok(())
+            }
+        };
+        Some(result)
+    }
+}
+
+impl Future for FutureVoid {
+    type Output = Result<(), crate::common::error::Error>;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let mut waker = self.waker.lock().unwrap();
+        if let Some(result) = self.try_get_result() {
+            // The future has completed. Remove the waker, if any, and return the result.
+            _ = waker.take();
+            Poll::Ready(result)
+        } else {
+            // The future has not completed yet. Do we need to register the callback?
+            match *waker {
+                Some(ref mut waker) => {
+                    // The callback has already been registered, just replace the waker.
+                    waker.clone_from(cx.waker());
+                }
+                None => {
+                    // Store the waker and drop the lock (in case the callback runs synchronously).
+                    *waker = Some(cx.waker().clone());
+                    drop(waker);
+
+                    // `user_data` will be cleaned up in `future_void_callback`.
+                    let user_data = Arc::into_raw(self.waker.clone()) as *mut ::libc::c_void;
+
+                    // SAFETY: `self.inner.as_ptr()` is a valid `aws_future_void` and this is the only callback we are registering.
+                    unsafe {
+                        aws_future_void_register_callback(self.inner.as_ptr(), Some(future_void_callback), user_data);
+                    }
+                }
+            }
+            Poll::Pending
+        }
+    }
+}
+
+/// Safety: Don't call this function directly, only called by the CRT as a callback.
+unsafe extern "C" fn future_void_callback(user_data: *mut ::libc::c_void) {
+    // Take ownership of the `Arc` in `user_data`.
+    let waker = Arc::from_raw(user_data as *mut Mutex<Option<Waker>>);
```

Comment:
I think it's best to distinguish:
* the `Arc` is there to share the waker (`Mutex<Option<Waker>>`, actually) between the struct and the callback. It uses the common `into_raw` - `from_raw` dance to pass a cloned instance into the "C" callback (see comments).
* The `Mutex` allows us to safely mutate it from the 2 sides. We make sure that the waker is updated on any `Pending` poll and removed once either the callback executes or poll discovers the future has completed, whichever happens first. (I'm adding a comment on the latter).

Reply:
I looked a bit more into the source code for `Arc`. It's just me misunderstanding the type. It generally sounds like it doesn't matter anyway, since we're only ever holding on to the `Arc` once.

My understanding of `from_raw`/`into_raw` is now this: `into_raw` simplify de-allocates the Arc without decrementing the reference counter. The agreement here is that it should be backed up with _one_ complementing `from_raw`, which will reconstruct an Arc retaining the original reference counter and does not increment it.

In our case, we don't really need the reference counting but this is our way of passing the struct through to the callback (avoiding additional allocations to the heap which would come with using `Box<..>`.

Reply:
Not quite:
* We have an `Arc` pointer in `FutureVoid::waker`, we clone it before registering the callback and then "leak" the clone into `user_data`. We rely on the guarantee that the callback will be invoked to justify the safety of the "leak".
* The memory pointed to by the `Arc` is shared by the struct and the callback (that's why we do need an `Arc`) and will only be de-allocated when both are dropped.